pkg/controller/deployment/deployment_controller.go

```diff
@@ -424,17 +433,64 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
-func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Deployment) error {
-	newRC, err := dc.getNewRC(deployment)
+// Rolling back to a revision; no-op if the toRevision is deployment's current revision
+func (dc *DeploymentController) rollback(deployment *extensions.Deployment, toRevision *int64) (*extensions.Deployment, error) {
+	newRC, allOldRCs, err := dc.getNewRCAndAllOldRCs(*deployment)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	allRCs := append(allOldRCs, newRC)
+	// If rollback revision is 0, rollback to the last revision
+	if *toRevision == 0 {
+		if *toRevision = lastRevision(allRCs); *toRevision == 0 {
+			// If we still can't find the last revision, gives up rollback
+			dc.emitRollbackWarningEvent(deployment, "DeploymentRollbackRevisionNotFound", "Unable to find last revision. Gives up rollback.")
```

Comment:
Didn't notice this until I reviewed #19835:

"Gives up rollback"doesn't use the same verb form as "Unable to find the last revision". Same issue below. 

Do we need additional text there at all? It seems obvious (to me, anyway) that if there is no previous revision, the rollback can't happen.

We could add something like "Rollback failed", "Rollback aborted", or somesuch, but I feel like they'd add more confusion rather than reduce it.